Usermanagement system

nico8034 opened the pull request.

OliverNymann committed:
UsermanagmentUI

nico8034 committed:
trying to fetch data to table

nico8034 committed:
Fixed the issue regarding cellvaluefactory property acess

nico8034 committed:
Updated sql read by role

OliverNymann committed:
Updated buttons

OliverNymann committed:
Merge branch 'usermanagement-manufacture' of https://github.com/Syhler/semester-project-2 into usermanagement-manufacture

nico8034 committed:
create users updated

Added create controller, database select for companies, create for user and differet useraccess regarding user role

nico8034 committed:
Created updateUser controller fixed arguments for create user

OliverNymann committed:
Delete added

OliverNymann committed:
Merge branch 'usermanagement-manufacture' of https://github.com/Syhler/semester-project-2 into usermanagement-manufacture

nico8034 committed:
update user finsihed

nico8034 committed:
Added stylingsheet and styling for usermangement

nico8034 committed:
updated design

nico8034 committed:
Added create, update delete company

added update of userinformation for signed in user

nico8034 committed:
Update currentuser updates currentuser UserEntiry

nico8034 committed:
Moved hashing of password to Domain layer, NEVER EVER DO DOMAINSTUFF IN PERSISTANCE AGAIN!!!!!

nico8034 committed:
Added language form usermangement

Usermangement
Company

nico8034 committed:
added russian for the memes

nico8034 committed:
Update UsermanagementController.java

OliverNymann committed:
Create/Update Language added

OliverNymann committed:
Merge branch 'usermanagement-manufacture' of https://github.com/Syhler/semester-project-2 into usermanagement-manufacture

nico8034 committed:
Added russian language and updated other languages

Syhler commented on src/main/java/org/example/domain/User.java:
```diff
@@ -1,21 +1,34 @@
-    private IPersistenceHandler persistenceHandler;
+    private IPersistenceHandler persistenceHandler = new PersistenceHandler();
+
+    public Long createUser(UserEntity userEntity,String password) {
+        var passwordSalt = PasswordHashing.generateSalt();
+        try {
+
+            var encryptedPassword = PasswordHashing.sha256(password,passwordSalt);
+            return persistenceHandler.user().createUser(userEntity, encryptedPassword, passwordSalt);
+
+        } catch (Exception e){
+            e.printStackTrace();
+        }
+
+        return 0L;
```
Lucas og jeg havde en samtale omkring "long", datatypen skal være med lille "l" også tror jeg ikke du har behøv for at sætte et "L" efter et tal :) - sorry for at fortæle dig noget andet

nico8034 replied:
Jeg har ændret Long til småt, forskellen er lige som Integer og interger fx, det er med stort er objekt versionen altså wrapper class

Syhler commented on src/main/java/org/example/entity/UserEntity.java:
```diff
@@ -3,16 +3,21 @@
     private String firstName;
     private String middleName;
     private String lastName;
+    private String fullName;
```
Er der nogen grund til at gemme "fullname" som en variable? Er det ikke muligt at return "firstName+" "+middleName+" "+lastName;" i "getFullName" metoden (linje 41-43) istedet?. Eller er der noget jeg ikke ser :)

nico8034 replied:
Du har ret, det kan jeg sagtens gøre i stedet

nico8034 replied:
Det viste sig faktisk ikke at være muligt grundet den måde cellfactory anvendes på tableview. Den tager dataen ud fra property altså som fullName, firstname osv og anvender get metoden til at få den.

Syhler replied:
Det iorden, jeg markere den her som resolve så :)

Syhler commented on src/main/java/org/example/entity/UserEntity.java:
```diff
@@ -25,12 +30,28 @@ public UserEntity(String title, String firstName, String middleName, String last
+        return fullName;
+    }
+
+    public void setFullName(String fullName) {
+        this.fullName = fullName;
     }
```
Bruger du nogensinde den her metode? hvis ja, hvorfor?

nico8034 replied:
Jeg har fjernet setFullName da den aldrig anvendes, men jeg er nød til beholde getFullName da cellfactory bruger den til at returnere variablen fullName

Syhler commented on src/main/java/org/example/persistence/PersistenceCompany.java:
```diff
@@ -0,0 +1,128 @@
+package org.example.persistence;
+
+import org.example.domain.password.PasswordHashing;
```
Laver et import fra "domain" laget, det må man ikke. Jeg tænker at det bare er en import som ikke bliver brugt, men du har glemt at fjerne :)

nico8034 replied:
Den blev ikke brugt, har fjernet den og de andre imports som ikke blev brugt

Syhler commented on src/main/java/org/example/persistence/PersistenceCompany.java:
```diff
@@ -0,0 +1,128 @@
+            e.printStackTrace();
+            return false;
+        }
+        return true;
```
Return 
 `preparedStatement.execute();`
så 
`return preparedStatement.execute();`

istedet for at `return true` på linje 83

Syhler commented on src/main/java/org/example/persistence/PersistenceCompany.java:
```diff
@@ -0,0 +1,128 @@
+                e.printStackTrace();
+                return false;
+            }
+            return true;
```
Return
preparedStatement.execute();
så
return preparedStatement.execute();

istedet for at return true på linje 61

Syhler commented on src/main/java/org/example/persistence/PersistenceCompany.java:
```diff
@@ -0,0 +1,128 @@
+            return 0L;
+        }
+
+    }
```
Der er 2 ting her:
1:
` Long id = companies.get(companies.size()-1).getId();`
`id++;`
behøves du ikke, da databasen gør det selv vedhjælp af serial :)

2:
Så har vi det samme med "Long", det skal være med lille "l", og så behøver du ikke at ende tallet med "L"

Syhler replied:
Du har fjernet det med Id'et. Men der mangler stadig det her med at ændre "Long" til "long"

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+                    }
+                };
+            }
+        };
```
Måske smide det her ind i sin egen metode, også få den metode til at return `Callback<ListView<CompanyEntity>, ListCell<CompanyEntity>>` istedet. Mere læseligt :)

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+            companyStage.initModality(Modality.WINDOW_MODAL);
+            companyStage.initOwner(((Node) event.getTarget()).getScene().getWindow());
+            companyStage.setResizable(false);
+            companyStage.showAndWait();
```
Her er der ingen grund til at bruge `showAndWait();` da du ikke venter på noget. `showAndWait();` gør det at programmet stopper din kode på den linje som metoden er kaldt, og venter med at køre den næste linje indtil den stage du har åbnet bliver lukket.

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+    public void createCompany(ActionEvent event) throws IOException {
+        CompanyEntity newCompany = new CompanyEntity(companyNameInput.getText());
+
+        Long id = domainHandler.company().createCompany(newCompany);
```
Igen det her med `Long`. brug `long` istedet

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+
+        } else {
+            setStatusText(newCompany.getName() + " wasn't created");
+        }
```
2 ting her.

1:
Når du laver `Long` om til `long`, kan du ikke tjekke om det er lig med null. `long` kan ikke være null, det kan `Long`. Fordi det er et object. :)

2:
I mangler languagehandler på linje 163: `setStatusText(newCompany.getName() + " was created");`
og på linje 167: `setStatusText(newCompany.getName() + " wasn't created");`

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+    @FXML
+    public void updateCompany(ActionEvent event) throws IOException {
+        CompanyEntity updatedCompany = new CompanyEntity(companyNameToUpdate.getText());
+        updatedCompany.setId(Long.parseLong(companyId.getText()));
```
Husk at ændre her også, når du ændrer `Long` til `long`

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+            companyEntities.add(updatedCompany);
+        } else {
+            setStatusText(updatedCompany.getName() + " wasn't updated");
+        }
```
2 ting her
1:
på linje 178 `if (domainHandler.company().updateCompany(updatedCompany)) {`
her ville jeg personligt kalde den her function en linje over også gemme boolean som bliver returnet af den metode i en variable, hvor den variables navn har noget der let beskriver hvad der bliver tjekket på.
f.eks. `if(companyWasUpdated)`... Håber du forstår hvad jeg prøver at sige :)

2:
Languagehandler igen på henholdvis linje 179; ` setStatusText(updatedCompany.getName() + " was updated");`
og linje 183: `setStatusText(updatedCompany.getName() + " wasn't updated");`

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+        } else {
+            setStatusText(selectedCompany.getName() + " wasn't deleted");
+        }
+    }
```
Det samme som kommentaren ovenover: 
henholdvis på linje 191 `if (domainHandler.company().deleteCompany(selectedCompany))`
192: `setStatusText(selectedCompany.getName() + " deleted");`
og 195: `setStatusText(selectedCompany.getName() + " wasn't deleted");`

Syhler commented on src/main/java/org/example/presentation/CreateUserController.java:
```diff
@@ -0,0 +1,225 @@
+    @FXML
+    private Label statusText;
+
+    public ObservableList<CompanyEntity> companyEntities = FXCollections.observableArrayList();
```
Nogen grund til den er public?

nico8034 replied:
Nej den er også blevet ændret til private

Syhler commented on src/main/java/org/example/presentation/CompanyController.java:
```diff
@@ -0,0 +1,220 @@
+    @FXML
+    private Label statusText;
+
+    public ObservableList<CompanyEntity> companyEntities = FXCollections.observableArrayList();
```
Nogen grund til den er public?

nico8034 replied:
nej det er en fejl, den er blevet ændret til private

Syhler commented on src/main/java/org/example/presentation/CreateUserController.java:
```diff
@@ -0,0 +1,225 @@
+     * @return currentUser
+     */
+    public UserEntity openCreateUser(ActionEvent event, int role) {
+        switch (role) {
+            case 1:
+                rolename = LanguageHandler.getText("admin");
+                break;
+            case 2:
+                rolename = LanguageHandler.getText("manufacture");
+                break;
+            case 3:
+                rolename = LanguageHandler.getText("producer");
+                break;
+            case 4:
+                rolename = LanguageHandler.getText("actor");
+                break;
+        }
```
Istedet for det her. Kan du kalde `Role.getRoleById();` hvor der så bliver returneret et Role object. For at få navnet på den rolle du får tilbage, kan du bare kalde `.toString();` efter :)
Det du har gang med her, ødelægger hele iden med enum, at vi nemlig undgår de her "magic" strings/numbers :)

nico8034 replied:
variablen rolename bliver blot brugt til title på det vindue der bliver åbnet, derfor jeg har brugt language handler, men jeg kan godt ændre numrene brugt i case.

Syhler replied:
Det fandme smart tænkt. Det overvejde jeg slet ikke xD

Syhler commented on src/main/java/org/example/presentation/CreateUserController.java:
```diff
@@ -0,0 +1,225 @@
+            FXMLLoader myLoader = App.getLoader("userManagementCreate");
+            createUserStage.setScene(new Scene(myLoader.load()));
+            CreateUserController createusercontrol = myLoader.getController();
+            createusercontrol.roleValue = role;
```
Her der sætter du en id af en rolle. Istedet burde du sætte et Role object. Nemlig det object du nu har fået returneret ved at kalde funktion `Role.getRoleById();`
Hvis du er i tvivl hvad jeg mener, skriver du bare :)

nico8034 replied:
Vi kan lige snakke om den her i morgen jeg tænkte nemlig på noget andet jeg kunne have gjort :)

Syhler commented on src/main/java/org/example/presentation/CreateUserController.java:
```diff
@@ -0,0 +1,225 @@
+        user.setCreatedByName(user.getCreatedBy().getFirstName());
+        user.setCompanyName(user.getCompany().getName());
+
+        Long userID = domainHandler.user().createUser(user, password.getText());
```
igen med `Long`  :)

Syhler commented on src/main/java/org/example/presentation/CreateUserController.java:
```diff
@@ -0,0 +1,225 @@
+
+        } else {
+            setStatusText("Something went wrong");
+        }
```
Her kan du fjerne "L" fra tallet.

Og du mangler langaugeHandler på linje 199: `setStatusText("Something went wrong");`

Syhler commented on src/main/java/org/example/presentation/DefaultController.java:
```diff
@@ -6,47 +6,93 @@
             login.setOnAction(this::logout);
+            profileNavigation.setVisible(true);
+
+            if (userEntity.getRole().getValue() < 4) {
+                usermanagementBtn.setVisible(true);
+            }
```
Er ikke sikker på jeg forstår?
Tror måske du prøver at tjekke om den som logger ind er en actor?
Hvis det er det du gør, kan det skrive som sådan her istedet:
`if(userEntity.getRole() == Role.Actor)`
Som er hele mening med enums. Her laver du et "magic" numbers. Noget man gerne ville ungå, så meget som muligt :)

nico8034 replied:
Ahh ja det har du helt ret i, aner faktisk ikke hvorfor jeg brugte et tal når vi alligevel har role enum til rådighed :D

Syhler commented on src/main/java/org/example/presentation/UpdateUserController.java:
```diff
@@ -0,0 +1,252 @@
+    @FXML
+    private Label statusText;
+
+    public ObservableList<CompanyEntity> companyEntities = FXCollections.observableArrayList();
```
Nogen grund til den er public?

nico8034 replied:
Samme som ved de andre, absolut ingen grund :B

Syhler commented on src/main/java/org/example/presentation/UpdateUserController.java:
```diff
@@ -0,0 +1,252 @@
+
+    @Override
+    public void initialize(URL url, ResourceBundle resourceBundle) {
+        Callback<ListView<CompanyEntity>, ListCell<CompanyEntity>> cellFactory = new Callback<ListView<CompanyEntity>, ListCell<CompanyEntity>>() {
+
+            @Override
+            public ListCell<CompanyEntity> call(ListView<CompanyEntity> l) {
+                return new ListCell<CompanyEntity>() {
+
+                    @Override
+                    protected void updateItem(CompanyEntity item, boolean empty) {
+                        super.updateItem(item, empty);
+                        if (item == null || empty) {
+                            setGraphic(null);
+                        } else {
+                            setText(item.getName());
+                        }
+                    }
+                };
+            }
+        };
```
Smid det her ind i en metode. Ligesom du gjorde tidligere :)

nico8034 replied:
Vi kigger lige på den her i morgen er ikke helt sikker på hvodan du vil have det

Syhler commented on src/main/java/org/example/presentation/UpdateUserController.java:
```diff
@@ -0,0 +1,252 @@
+            case 4:
+                rolename = LanguageHandler.getText("actor");
+                break;
+        }
```
Her har vi det igen. Ændre det, som det jeg beskrev i den anden controller, som næsten har det samme kode.

Syhler commented on src/main/java/org/example/presentation/UpdateUserController.java:
```diff
@@ -0,0 +1,252 @@
+            FXMLLoader myLoader = App.getLoader("userManagementUpdate");
+            updateUserStage.setScene(new Scene(myLoader.load()));
+            UpdateUserController updateUserController = myLoader.getController();
+            updateUserController.roleValue = role;
```
Brug enum her også :)

Syhler commented on src/main/java/org/example/presentation/UpdateUserController.java:
```diff
@@ -0,0 +1,252 @@
+            closeDialog(event);
+
+        } else {
+            setStatusText("Something went wrong");
```
Languagehandler :)

Syhler commented on src/main/java/org/example/presentation/UsermanagementController.java:
```diff
@@ -1,24 +1,231 @@
+    private TableColumn<UserEntity, String> col_created;
+
+
+    public ObservableList<UserEntity> userList = FXCollections.observableArrayList();
```
Nogen grund til den er public?

Syhler commented on src/main/java/org/example/presentation/UsermanagementController.java:
```diff
@@ -1,24 +1,231 @@
+
+    public ObservableList<UserEntity> userList = FXCollections.observableArrayList();
+
+    public int roleTap = 0;
```
Hvad er det her? hvad er "roleTap"?

nico8034 replied:
Har forklaret og omkskrevet i din commet lidt længere nede

Syhler commented on src/main/java/org/example/presentation/UsermanagementController.java:
```diff
@@ -1,24 +1,231 @@
+
+        companyAddToggle.setVisible(false);
+
+        switch (CurrentUser.getInstance().getUserEntity().getRole().getValue()) {
+            case 1:
+                displayAdmins.fire();
+                displayAdmins.setSelected(true);
+                companyAddToggle.setVisible(true);
+                break;
+            case 2:
+                displayAdmins.setVisible(false);
+                displayManufactures.setVisible(false);
+                displayProducers.fire();
+                displayProducers.setSelected(true);
+                break;
+            case 3:
+                displayAdmins.setVisible(false);
+                displayManufactures.setVisible(false);
+                displayProducers.setVisible(false);
+                displayActors.fire();
+                displayActors.setSelected(true);
+                break;
+            case 4:
+                displayAdmins.setVisible(false);
+                displayManufactures.setVisible(false);
+                displayProducers.setVisible(false);
+                displayActors.setVisible(false);
+                break;
+        }
```
Fjern `.getValue()` også udskifte tallene i "case" til de forskellige roller. Så f.eks.
`case Role.Admin:`

nico8034 replied:
Har omskrevet den og fjernet getValue() så den nu tjekker på Enum konstanterne Admin, Manufacture osv

Syhler commented on src/main/java/org/example/presentation/UsermanagementController.java:
```diff
@@ -1,24 +1,231 @@
+
+        if (displayByRole == displayAdmins) {
+            userList.addAll(domainHandler.user().getUserByRole(Role.Admin));
+            roleTap = 1;
+            displayManufactures.setSelected(false);
+            displayProducers.setSelected(false);
+            displayActors.setSelected(false);
+        }
+        if (displayByRole == displayManufactures) {
+            userList.addAll(domainHandler.user().getUserByRole(Role.Manufacture));
+            roleTap = 2;
+            displayAdmins.setSelected(false);
+            displayProducers.setSelected(false);
+            displayActors.setSelected(false);
+        }
+        if (displayByRole == displayProducers) {
+            userList.addAll(domainHandler.user().getUserByRole(Role.Producer));
+            roleTap = 3;
+            displayManufactures.setSelected(false);
+            displayAdmins.setSelected(false);
+            displayActors.setSelected(false);
+        }
+        if (displayByRole == displayActors) {
+            userList.addAll(domainHandler.user().getUserByRole(Role.Actor));
+            roleTap = 4;
```
RoleTap?
Hvis jeg forstår det rigtig, så er det noget med rollerne at gøre?`
Hvis det er det, så gem dem i en Enum altså Role. også kan du sætte den ligmed en rolle som f.eks.
`role = Role.Admin`

nico8034 replied:
Har lavet dem om så jeg bruger funktionen getValue() som allerede var skrevet i Enum klassen, roleTap = Role.Admin.getValue();  ved ikke om det var sådan du gerne ville have det. Roletap er en værdi jeg sender med over til createUserController så jeg ved hvilken rolle de skal oprettes med.

Syhler replied:
Lad os tage den her i morgen. Nemmere at vise hvad jeg mener :)

Syhler reviewed (commented):
Også, mangler i rigtig mange kommentar :)

nico8034 reviewed (commented):
Så skulle alle ting med dine comments gerne være rettet, jeg har svaret tilbage med en comment på dem alle. Har også fjernet id fra createUser, så databasen håndtere det, det mener jeg bare ikke den kunne på daværende tidspunkt, men jeg har testet og det virker :)

nico8034 committed:
Small changes and optimizations of the code

nico8034 committed:
removed generated ID from createCompany and some outcommented code

nico8034 committed:
more fixes and changes

nico8034 committed:
Update UserEntity.java

nico8034 committed:
minor fixes

nico8034 committed:
Update english.xml

nico8034 committed:
Added utility package

moved form validation, setStagetitle and cellfactory callback into usermanagementUtility class

nico8034 committed:
Added comments for methods

Syhler committed:
Did some stuff yes yes big stuff
